### PR TITLE
object detection examples notebook

### PR DESCRIPTION
**Pull request by aditya1503**

### CODE REVIEW

**Review by ulya-tkch — APPROVED** [review]
Approved after addressing PR comments on wording. Great job!

**Comment by jwmueller** [comment]
reminder for myself: add to table of contents right before merge

File: object_detection/detectron2_training.ipynb [hunk 1]
```diff
@@ -0,0 +1,243 @@
+{
+ "cells": [
```
Comment (jwmueller) [inline object_detection/detectron2_training.ipynb:2]: need requirements.txt file. Make sure the first line of requirements file says:

\# This notebook currently requires the latest developer version of cleanlab: pip install git+https://github.com/cleanlab/cleanlab.git

Comment (jwmueller) [reply object_detection/detectron2_training.ipynb:2]: need a README inside the object_detection folder that in 1-line describes what is difference between the two notebooks

File: object_detection/detectron2_training-kfold.ipynb [hunk 2]
```diff
@@ -0,0 +1,337 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Training an Object Detection model using Detectron2 with Stratified K-Fold\n",
```
Comment (jwmueller) [inline object_detection/detectron2_training-kfold.ipynb]:
```suggestion
    "# Fitting a (Detectron2) Object Detection Model via K-Fold Cross-Validation to produce Out-of-sample Predictions for every Image in a Dataset \n",
```

File: object_detection/detectron2_training.ipynb [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Training an Object Detection model using Detectron2\n",
+    "\n",
+    "This 5-minute quickstart tutorial demonstrates how to train Detectron2 on object detection datasets. In this dataset each example contains a bounding box and a class label surrounding a physical object within an image scene. Using this labeled data, we train a model to predict classes of objects in an image and their physical locations.\n",
```
Comment (jwmueller) [inline object_detection/detectron2_training.ipynb]:
```suggestion
    "This notebook demonstrates how to train a [Detectron2](https://github.com/facebookresearch/detectron2/) model on object detection datasets and produce predictions required to run cleanlab's tutorial on detecting label errors in object detection data.  Note that this notebook fits the model to an entire training set and produces predictions on a held-out validation set. Thus these predictions are only *out-of-sample* for the validation data, and should ideally *only* be used to find mislabeled images amongst the validation set. To instead find mislabeled images amongst an entire dataset, see the analogous notebook in this folder which uses K-fold cross-validation to produce out-of-sample predictions for every image in the dataset.\n",
    "\n",
    "In object detection data, each image is annotated with multiple bounding boxes.  Each bounding box surrounds a physical object within an image scene, and is annotated with a given class label. Using this labeled data, we train a model to predict the locations and classes of objects in an image. The trained model can subsequently be used to identify mislabeled images, which when corrected, allow you to train an even better model without changing your training code! \n",
```

File: object_detection/detectron2_training.ipynb [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    "\n",
+    "This notebook demonstrates how to train a state-of-the-art object detection model using Detectron2 and use it to produce `pred_probs`, which will help detect label errors in Object detection datasets. \n",
+    "\n",
+    "To identify label errors, we provide an [example](https://github.com/cleanlab/examples/) notebook on our GitHub repository on [Finding Label Errors in Object Detection Datasets](https://github.com/cleanlab/cleanlab/blob/master/docs/source/tutorials/object_detection.ipynb). Here we fit a state-of-the-art neural network initialized from a pretrained [X-101](https://github.com/facebookresearch/detectron2/blob/main/MODEL_ZOO.md#imagenet-pretrained-models) network backbone."
```
Comment (jwmueller) [inline object_detection/detectron2_training.ipynb]:
```suggestion
    "Here we will fit a state-of-the-art neural network trained starting from a pretrained [X-101](https://github.com/facebookresearch/detectron2/blob/main/MODEL_ZOO.md#imagenet-pretrained-models) network backbone. First let's import the required packages and download the data."
```

File: object_detection/detectron2_training.ipynb [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We define the configuration settings for training an object detection model using Detectron2. The model architecture used in this example is \"faster_rcnn_X_101_32x8d_FPN_3x\" from the COCO-Detection model zoo. The training data is specified by the \"my_dataset_train\" dataset and validation data is specified by the \"my_dataset_val\" dataset which refer to COCO2017 train and val containing only the subset of labels specified before.\n",
+    "\n",
+    "The number of worker threads is set to 2 and the batch size is set to 2.\n",
+    "The learning rate and maximum number of iterations are also specified. The model is initialized from the COCO-Detection model zoo and the output directory for the trained model is created. Finally, the configuration is passed to the DefaultTrainer class for training the object detection model.\n",
```
Comment (jwmueller) [inline object_detection/detectron2_training.ipynb]:
```suggestion
    "Here the number of worker threads is set to 2 and the batch size is set to 2. The learning rate and maximum number of iterations are also specified. You'll want to tinker with these values to get the best performance for your own data.\n",
    "The model is initialized from the COCO-Detection model zoo and the output directory for the trained model is created. Finally, this configuration is passed to the DefaultTrainer class for training the object detection model.\n",
```

File: object_detection/detectron2_training.ipynb [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+   "outputs": [],
+   "source": [
+    "cfg = get_cfg()\n",
+    "cfg.merge_from_file(model_zoo.get_config_file(\"COCO-Detection/faster_rcnn_X_101_32x8d_FPN_3x.yaml\"))\n",
+    "cfg.DATASETS.TRAIN = (\"my_dataset_train\",)\n",
+    "cfg.DATASETS.TEST = (\"my_dataset_val\",)\n",
+    "cfg.DATALOADER.NUM_WORKERS = 2\n",
+    "cfg.MODEL.WEIGHTS = model_zoo.get_checkpoint_url(\"COCO-Detection/faster_rcnn_X_101_32x8d_FPN_3x.yaml\")  # Let training initialize from model zoo\n",
+    "cfg.SOLVER.IMS_PER_BATCH = 2  # This is the real \"batch size\" commonly known to deep learning people\n",
+    "cfg.SOLVER.BASE_LR = 0.00025  # pick a good LR\n",
```
Comment (jwmueller) [inline object_detection/detectron2_training.ipynb]:
```suggestion
    "cfg.SOLVER.BASE_LR = 0.00025  # IMPORTANT: pick a good Learning Rate for your dataset \n",
```

File: object_detection/detectron2_training.ipynb [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    "cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, \"model_final.pth\")  # path to the model we just trained\n",
+    "cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.7   # set a custom testing threshold\n",
+    "predictor = DefaultPredictor(cfg)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The given code block implements a function \"format_detectron2_predictions\" that converts the output of Detectron2 to a format that can be used by Cleanlab for identifying label errors. The function accepts the predicted instances and the number of classes as inputs. It processes the predicted bounding boxes and scores for each instance, and outputs a list of numpy arrays containing the bounding boxes and scores for each class."
```
Comment (jwmueller) [inline object_detection/detectron2_training.ipynb]:
```suggestion
    "The code below defines a function \"format_detectron2_predictions\" to convert the prediction output of Detectron2 to a format that can be used by Cleanlab for identifying label errors. This function accepts the predicted object instances and the number of classes as inputs. It processes the predicted bounding boxes and prediction-confidence for each instance, and outputs a list of numpy arrays containing the bounding boxes and prediction-confidence for each class."
```

Comment (jwmueller) [reply object_detection/detectron2_training.ipynb]: is it for each class? I thought it was just for the predicted (i.e. most likely) class

File: object_detection/detectron2_training.ipynb [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    "        res2.append(q)\n",
+    "    return res2\n",
+    "        \n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To perform inference and testing on the tutorial notebook linked here, we utilize a limited portion of the validation set of COCO 2017. To find label errors on this subset, please run our tutorial notebook on [Finding Label Errors in Object Detection Datasets](https://github.com/cleanlab/cleanlab/blob/master/docs/source/tutorials/object_detection.ipynb) to find label errors."
```
Comment (jwmueller) [inline object_detection/detectron2_training.ipynb]:
```suggestion
    "To perform inference and testing on the tutorial notebook linked here, we utilize a limited portion of the validation set of COCO 2017. To find label errors in this subset, please run our [tutorial](https://docs.cleanlab.ai/stable/index.html) notebook on [Finding Label Errors in Object Detection Datasets](https://github.com/cleanlab/cleanlab/blob/master/docs/source/tutorials/object_detection.ipynb)."
```

File: object_detection/detectron2_training-kfold.ipynb [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+    "\n",
+    "This notebook demonstrates how to train a state-of-the-art object detection model using Detectron2 and use it to produce `pred_probs`, which will help detect label errors in Object detection datasets.<b> In addition, we will utilize Stratified K-Fold with a seed to produce out-of-distribution predicted probabilities for the training data.</b> It is important to not provide predictions on the same datapoints used to train the model, as these will be overfitted and unsuitable for finding label issues.\n",
+    "\n",
+    "To find label errors on this subset, please run our tutorial notebook on [Finding Label Errors in Object Detection Datasets](https://github.com/cleanlab/cleanlab/blob/master/docs/source/tutorials/object_detection.ipynb) to find label errors with this data.\" to the end of this notebook like your other training notebook. Even though it is not the same dataset, the results could be evaluated in the same way.\n"
```
Comment (jwmueller) [inline object_detection/detectron2_training-kfold.ipynb]:
```suggestion
    "This notebook demonstrates how to fit a [Detectron2](https://github.com/facebookresearch/detectron2/) model on object detection datasets via *cross-validation*, and produce *out-of-sample* predictions required to run cleanlab for detecting label errors in such data.  Out-of-sample predictions (from a copy of the model that was never trained on the image it is producing a prediction for) are less subject to overfitting and can more accurately detect label issues, but are harder to obtain for every image in an entire dataset unless we employ K-fold cross-validation as demonstrated here.\n",
    "\n",
    "Thus run this notebook if you wish to find every mislabeled image an a dataset. Once you have the out-of-sample predictions produced from this notebook, you can directly run the cleanlab [tutorial](https://docs.cleanlab.ai/stable/index.html) notebook on [Finding Label Errors in Object Detection Datasets](https://github.com/cleanlab/cleanlab/blob/master/docs/source/tutorials/object_detection.ipynb). Even though the cleanlab tutorial is focused on finding issues in only the validation subset of COCO, you can easily update the tutorial to run on the full dataset for which we obtain predictions via cross-validation in this notebook.\n",
    "\n",
    "In object detection data, each image is annotated with multiple bounding boxes.  Each bounding box surrounds a physical object within an image scene, and is annotated with a given class label. Using this labeled data, we train a model to predict the locations and classes of objects in an image. The trained model can subsequently be used to identify mislabeled images, which when corrected, allow you to train an even better model without changing your training code! \n",
    "\n",
    "Here we fit a state-of-the-art neural network trained starting from a pretrained [X-101](https://github.com/facebookresearch/detectron2/blob/main/MODEL_ZOO.md#imagenet-pretrained-models) network backbone. First let's import the required packages and download the data."
```

File: object_detection/detectron2_training-kfold.ipynb [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This code performs K-Fold cross-validation on a dataset in the COCO format.\n",
+    "* It splits the dataset into multiple folds and creates separate training and test data for each fold. The code loads a JSON file containing the dataset annotations, extracts the image IDs and maps them to file names. \n",
+    "* It then defines a function to split the data based on the fold indices. The function creates separate training and test data dictionaries, preserving common metadata. It iterates over the images and annotations, assigning them to the appropriate data split based on the image ID. \n",
+    "*Finally, it saves the training and test data as separate JSON files and prints information about the data in each fold, such as the number of images and annotations."
```
Comment (jwmueller) [inline object_detection/detectron2_training-kfold.ipynb]:
```suggestion
    "* Finally we save the training and test data as separate JSON files and print information about each fold."
```

File: object_detection/detectron2_training-kfold.ipynb [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "from sklearn.model_selection import KFold\n",
+    "from collections import defaultdict\n",
+    "NUM_FOLDS = 5\n",
```
Comment (jwmueller) [inline object_detection/detectron2_training-kfold.ipynb]:
```suggestion
    "NUM_FOLDS = 5  # Reduce this to get faster results but potentially less accurate predictions \n",
```

File: object_detection/detectron2_training-kfold.ipynb [hunk 12]
```diff
@@ -0,0 +1,337 @@
+{
```
Comment (jwmueller) [inline object_detection/detectron2_training-kfold.ipynb:1]: where appropriate also incorporate my suggested changes for the other notebook into this one

File: object_detection/detectron2_training.ipynb (hunk 7 above)
Comment (aditya1503) [reply object_detection/detectron2_training.ipynb]: yes most confident class

Comment (jwmueller) [reply object_detection/detectron2_training.ipynb]: then please delete: "for each class" at the end

File: object_detection/requirements.txt [hunk 13]
```diff
@@ -0,0 +1,5 @@
+numpy==1.22.0
+detectron2==0.6
+torch==1.110+cu113
+pandas==1.5.0
+# Note: these examples only work with developer of cleanlab until v2.4 is released
```
Comment (jwmueller) [inline object_detection/requirements.txt]:
```suggestion
# Note: for now, these examples only work with the developer version of cleanlab: `pip install git+https://github.com/cleanlab/cleanlab.git`
```

**Review by jwmueller — APPROVED** [review]
LGTM

**Pull request merged**